File: Swift-C++-Testsuite/test/enumerations/Swift/EnumAssociatedValues.swift
// You can define Swift enumerations to store associated
// values of any given type, and the value types can be
// different for each case of the enumeration if needed.
// Enumerations similar to these are known as discriminated
// unions, tagged unions, or variants in other programming languages.

enum Barcode {
    case upc(Int, Int, Int, Int)
    case qrCode(String)
}
@main
struct MyMain {

    static func main() {

        var productBarcode = Barcode.upc(8, 85909, 51226, 3)
        let first = strangeSum(productBarcode)
        productBarcode = .qrCode("ABCDEFGHIJKLMNOP")
        let second = strangeSum(productBarcode)
    }

    static func strangeSum(_ code: Barcode) -> Int {
        var x = 0
        switch code {
        case .upc(let numberSystem, let manufacturer, let product, let check):
            x = numberSystem + manufacturer + product + check
        case .qrCode(let productCode):
            x = -1
        }
        return x
    }
}

File: Swift-C++-Testsuite/test/enumerations/C++/EnumAssociatedValues.cpp
#include<variant>
#include<string>
#include<tuple>
#include <iostream>
using namespace std;

int strangeSum(std::variant<std::tuple<int, int, int, int>, std::string> code) {
	// Initial value set to 0   
    int x = 0;

    // Check the type of the variable
    if(auto number_system = std::get_if<0>(&code)){
        x = std::get<0>(*number_system) + std::get<1>(*number_system) + std::get<2>(*number_system) + std::get<3>(*number_system);
    } else if(auto product_code = std::get_if<1>(&code)) {
        x = -1;
    }

    return x;
}

int main(){
	// Variable for main
    std::variant<std::tuple<int, int, int, int>, std::string> productBarcode;
	
    productBarcode = std::make_tuple(8, 85909, 51226, 3);
	int first = strangeSum(productBarcode);

	productBarcode = std::string("ABCDEFGHIJKLMNOP");
	int second = strangeSum(productBarcode);

	return 0;
}